CorsixTH/Src/th_gfx_sdl.cpp

```diff
@@ -722,7 +724,69 @@ void render_target::set_cursor_position(int iX, int iY) {
   cursor_y = iY;
 }
 
-bool render_target::take_screenshot(const char* sFile) {
+namespace {
+
+bool write_rgb_png(int width, int height, png_bytep pixels, int pitch,
+                   const char* file_path) {
+  // Open the file for writing
+  FILE* fp = std::fopen(file_path, "wb");
+  if (fp == nullptr) {
+    return false;
+  }
+
+  // Create PNG writer and info data structures.
+  png_structp png_write_data =
+      png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
+  if (png_write_data == nullptr) {
+    std::fclose(fp);
+    return false;
+  }
+  png_infop info_write_data = png_create_info_struct(png_write_data);
+  if (info_write_data == nullptr) {
+    png_destroy_write_struct(&png_write_data, nullptr);
+    std::fclose(fp);
+    return false;
+  }
+
+  // Setup error handling.
+  if (setjmp(png_jmpbuf(png_write_data))) {
```

Comment:
To review further, there can be complications with jmp in c++.

Reply:
So the caveat here is that it's really easy to trigger undefined behaviour, if any non-trivial c++ objects would have been destroyed by the unwinding of the longjmp.

I don't think that is the case here, but I find it hard to reason about longjmp. I think you may be leaking the array below? You also couldn't turn that c array into a c++ one because that would definitely trigger the ub.

I think libpng offers alternative error handling.

Reply:
Hmm, yes good spot.

Row pointers allocation should be moved then indeed.

Alternative error handling seems to need a custom compile:
```
If you would rather avoid the complexity of setjmp/longjmp issues,
you can compile libpng with PNG_NO_SETJMP, in which case
errors will result in a call to PNG_ABORT() which defaults to abort().

You can #define PNG_ABORT() to a function that does something
more useful than abort(), as long as your function does not
return.
```

Reply:
After moving some of the stuff here I found new leaks. To fix once and for all, I centralized memory management at the top and bottom of the function.